src/wallet/wallet.cpp

```diff
@@ -1775,9 +1839,21 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
             WalletLogPrintf("Still rescanning. At block %d. Progress=%f\n", block_height, progress_current);
         }
 
+        bool skip_block = false;
+        if (fast_rescan_filter) {
+            fast_rescan_filter->UpdateIfNeeded();
+            auto match_result = fast_rescan_filter->MatchesBlock(block_hash);
+            if (match_result.has_value() && !match_result.value()) {
```

Comment:
I don't think that's logically equivalent. 
* "we have a result, and the result is negative" (PR), versus
* "we have a result, and the result is positive" (your suggestion)

(according to https://en.cppreference.com/w/cpp/utility/optional/value_or: "Returns the **contained value** if *this has a value, otherwise returns default_value.")

Reply:
You're right, I should have inversed it. This should be equivalent though:

```suggestion
            if (!match_result.value_or(true)) {
```

Reply:
While I agree that this is now logically equivalent, I doubt that it's more readable.

Reply:
Maybe it's me, but for some reason I'm often confused by `!` in general, including in the second hand of the original version (as per my initial wrong suggestion, apparently). What about `match_result.value_or(true) == false)`? I think that's concise and clear. I'm just not a fan of "re-implementing" the `value_or()` when it's so common.

Don't want to bikeshed over this obviously, so as you see fit - no further comment.

Reply:
I'm still not convinced, as I think there is more mental effort needed for a reader to grasp exactly what the "block skip condition" is. "the filter matching routine returned a result, _and_ that result is negative" is expressed more closely with the PR variant IMHO. Will still leave the discussion open for a while in case anyone wants to chime in. Maybe it's more a matter of taste and I have a bad one 😛

Reply:
As we are not using the returned optional value (which, if nullopt denotes that we don't have the filter index for it) and we only want to know if the set matches the block filter, could change `MatchesBlock` to return a simple boolean instead.

Something like:
```c++
std::optional<bool> MatchesBlock(const uint256& block_hash) const
{
    auto matches{m_wallet.chain().blockFilterMatchesAny(BlockFilterType::BASIC, block_hash, m_filter_set)};
    return matches && *matches;
}
```

And then this specific line would just be:
```suggestion
            if (match_result) {
```

Reply:
@furszy: Unfortunately that doesn't work. Remember that the condition for skipping a block if the filter does _not_ match  the block. If we just use `!match_result` as condition (I guess in your suggested change your just forgot the negation operator `!`), then we would also skip a block if the match function didn't return a result (e.g. because the block filter is not available), which is incorrect.

Reply:
I guess in your suggested change your just forgot the negation operator !

it was a `!match_result`, yeah.

> then we would also skip a block if the match function didn't return a result (e.g. because the block filter is not available), which is incorrect.

That is guarded by the `fast_rescan_filter` object existence (before create the object, you check if the filter index exist). If there isn't a filter index, then the object is never created, there by this line is never called.

As block filters are created with blocks (on the same connection signal), we shouldn't need to check for filter existence at this point of the code.

Still.. there is still a tiny window of time for a race condition.. when a block gets stored and we are still creating the block filter..
So.. better to check the filter existence on every loop. And forget all what I just said :p

Reply:
A less controversial idea to improve readability: I think if the name represents what true/false means, it becomes clearer: for example `matches_block` instead of `match_result`. Also in this case, I think dereferencing is clearer than using `.value()`

<details>
<summary>git diff</summary>

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index d3f1fb418..2c2f80cfa 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1846,8 +1846,8 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
         bool skip_block{false};
         if (fast_rescan_filter) {
             fast_rescan_filter->UpdateIfNeeded();
-            auto match_result{fast_rescan_filter->MatchesBlock(block_hash)};
-            if (match_result.has_value() && !match_result.value()) {
+            auto matches_block{fast_rescan_filter->MatchesBlock(block_hash)};
+            if (matches_block.has_value() && !*matches_block) {
                 result.last_scanned_block = block_hash;
                 result.last_scanned_height = block_height;
                 skip_block = true;
///

</details>

Reply:
Thanks, done 👌  I agree this is more readable than the previous version.

Reply:
The next commit aa613697e9c0b9005d411562bfdf8bd727c0b02f makes things more readable anyway because both the `true` and `false` case are handled for logging.